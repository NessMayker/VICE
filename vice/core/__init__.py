
# __all__ = ["integrator", "output"]

import _wrapper
import _data_management
import _globals

__all__ = []

for i in range(len(_wrapper.__all__)):
	__all__.append(_wrapper.__all__[i])
for i in range(len(_data_management.__all__)):
	__all__.append(_data_management.__all__[i])
for i in range(len(_globals.__all__)):
	__all__.append(_globals.__all__[i])

from _wrapper import *
from _data_management import *
from _globals import *

