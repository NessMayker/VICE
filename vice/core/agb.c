
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <math.h>
#include "specs.h"
#include "stars.h"
#include "enrichment.h"

static double get_AGB_yield(INTEGRATION run, int index, long time_index, 
	double zto);
// static void setup_single_AGB_grid(element *e, double **grid, double *times, 
// 	long num_times);
static int *get_bounds(double *arr, double value, int length);
static double interpolate(double mz1, double mzto, double mz2, double y1, 
	double y2);



extern double m_AGB(INTEGRATION run, MODEL m, int index) {

	if (run.timestep == 0) {
		return 0;
	} else {
		double mass = 0, solar = 0;
		long i;
		int j;
		for (j = 0; j < run.num_elements; j++) {
			solar += run.elements[j].solar;
		}
		for (i = 0l; i < run.timestep; i++) {
			double Z = 0;
			for (j = 0; j < run.num_elements; j++) {
				Z += run.Zall[j][run.timestep - i];
			}
			Z *= m.Z_solar / solar;
			// double y = get_AGB_yield(run, index, i, Z);
			mass += (get_AGB_yield(run, index, i, Z) * 
				run.mdotstar[run.timestep - i] * run.dt * (m.H[i] - 
					m.H[i + 1]));
		}
		return mass;
	}

}

/*
Returns the effective fractional AGB yield of the given element at the 
lookback time time_index timesteps ago.

Args:
=====
run:		The integration struct for the current execution
index:		The index of the element being enriched
time_index:	The number of timesteps ago that the stellar population formed
zto:		The metallicity of the stellar population that formed.
*/
static double get_AGB_yield(INTEGRATION run, int index, long time_index, 
	double zto) {

	int *bounds = get_bounds(run.elements[index].agb_z, zto, 
		run.elements[index].num_agb_z);
	if (bounds[0] == -1) {
		/*
		These lines used to tie the yields down to 0 at z = 0. This is probably 
		a good assumption for s-process elements like strontium, but not for 
		others like carbon, where the yields decrease with increasing 
		metallicity 

		double yield = interpolate(0, zto, 
			run.elements[index].agb_z[bounds[1]], 
			0, run.elements[index].agb_grid[time_index][bounds[1]]);
		free(bound);
		return yield;
		*/

		free(bounds);
		return interpolate(run.elements[index].agb_z[0], zto, 
			run.elements[index].agb_z[1], 
			run.elements[index].agb_grid[time_index][0], 
			run.elements[index].agb_grid[time_index][1]);
	} else {
		if (bounds[1] == run.elements[index].num_agb_z) {
			bounds[0]--;
			bounds[1]--;
		} else {}
		double yield = interpolate(run.elements[index].agb_z[bounds[0]], zto, 
			run.elements[index].agb_z[bounds[1]], 
			run.elements[index].agb_grid[time_index][bounds[0]], 
			run.elements[index].agb_grid[time_index][bounds[1]]
		);
		free(bounds);
		return yield;
	}

}

#if 0
Sets up the AGB yield grids for each element.

Args:
=====
run:		The integration struct for the current execution
grids:		The grids sampled at various inital stellar masses and 
			metallicities
times:		The times that the integration will evaluate at
num_times:	The number of elements in the array times

extern int setup_AGB_grids(integration *run, double ***grids, double *times, 
	long num_times) {

	int i;
	for (i = 0; i < (*run).num_elements; i++) {
		setup_single_AGB_grid(&(*run).elements[i], grids[i], times, num_times);
	}
	return 0;

}
#endif

/*
Sets up the yield grid for a single element.

Args:
=====
e:			A pointer to the element struct to hold the grid
grid:		The grid sampled at various masses and metallicities for this 
			element
times:		The times that the integration will evaluate at
num_times:	The number of elements in the array times.
*/
extern void setup_single_AGB_grid(ELEMENT *e, double **grid, double *times, 
	long num_times) {

	/*
	The grid is composed of an array of yields for each turnoff mass that the 
	integration will see given the times that it will evaluate at. It is then 
	sampled at the metallicities within the grid and interpolated between 
	at each timestep.

	Within each element's grid is a row of yields at the corresponding turn 
	off mass for each metallicity sampled on the grid.
	*/
	long i;
	int j;
	e -> agb_grid = (double **) malloc (num_times * sizeof(double *));
	for (i = 0l; i < num_times; i++) {
		e -> agb_grid[i] = (double *) malloc (e -> num_agb_z * sizeof(double));
		double mto = m_turnoff(times[i]);
		int *bounds = get_bounds(e -> agb_m, mto, e -> num_agb_m);
		for (j = 0; j < (*e).num_agb_z; j++) {
			if (mto > 8) {
				/* Tie it doen to 0 at m > 8 Msun */
				e -> agb_grid[i][j] = 0;
			} else if (bounds[0] == -1) {
				/* Tie it down to y = 0 at m = 0 */
				e -> agb_grid[i][j] = interpolate(0, mto, e -> agb_m[0], 
					0, grid[0][j]);
			} else if (bounds[1] == (*e).num_agb_m) {
				/* Tie it down to y = 0 at 8 Msun */
				e -> agb_grid[i][j] = interpolate((*e).agb_m[bounds[0]], mto, 
					8, grid[bounds[0]][j], 0);
			} else {
				/* Interpolate normally otherwise */ 
				e -> agb_grid[i][j] = interpolate(e -> agb_m[bounds[0]], mto, 
					e -> agb_m[bounds[1]], grid[bounds[0]][j], 
					grid[bounds[1]][j]);
			}
		}
		free(bounds);
	}

}

/*
Returns a pointer to two integers. The first is the index of the number in the 
array that is the largest number smaller than the value. The second is the 
index of the number in the array the is the smaller number larger than the 
value. In other words, they're the bounding indeces that can be used for 
interpolation.

Args:
=====
arr:		The array to find bounding indeces within
value:		The value to find bounding indeces for
length:		The number of elements in the array arr
*/
static int *get_bounds(double *arr, double value, int length) {

	int *indeces = (int *) malloc (2 * sizeof(int));
	if (value < arr[0]) {
		indeces[0] = -1;
		indeces[1] = 0;
		return indeces;
	} else {
		int i;
		for (i = 0; i < length; i++) {
			if (arr[i] >= value) {
				indeces[0] = i - 1;
				indeces[1] = i;
				return indeces;
			} else {
				continue;
			}
		}
		indeces[0] = length - 1;
		indeces[1] = length;
		return indeces;
	}

}

/*
Interpolates between two yields in either mass or metallicity space.

Args:
=====
mz1:		The lower mass/metallicity to interpolate from
mzto:		The mass/metallicity dependent yield being approximated
mz2:		The upper mass/metallicity to interpolate from
y1:			The yield at mz1
y2:			The yield at mz2
*/
static double interpolate(double mz1, double mzto, double mz2, double y1, 
	double y2) {

	return y1 + (y2 - y1) / (mz2 - mz1) * (mzto - mz1);

}

