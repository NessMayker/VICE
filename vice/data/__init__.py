
from __future__ import absolute_import

__all__ = [str("agb_yield_grid"), str("fractional_cc_yield")]

from ._agb_yields import yield_grid as agb_yield_grid
from ._ccsne_yields import fractional_cc_yield

