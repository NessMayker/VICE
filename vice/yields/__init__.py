"""
This file, included with the VICE package, is protected under the terms of the 
associated MIT License, and any use or redistribution of this file in original 
or altered form is subject to the copyright terms therein. 
"""

from __future__ import absolute_import
from . import agb 
from . import ccsne 
from . import sneia 

__all__ = ["agb", "ccsne", "sneia"] 

del absolute_import 
