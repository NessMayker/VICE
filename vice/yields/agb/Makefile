
.PHONY: clean

clean: 
	rm -f grid.c 

