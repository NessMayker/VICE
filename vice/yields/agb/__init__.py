# This file, included with the VICE package, is protected under the terms of the 
# associated MIT License, and any use or redistribution of this file in original 
# or altered form is subject to the copyright terms therein. 

"""
Asymptotic Giant Branch Star Nucleosynthetic Yield Tools 
========================================================
In the current version of VICE, nucleosynthetic yields from AGB stars allows 
users to select between the tables published in the Karakas (2010) and the 
Cristallo et al. (2011) studies. 

Included Features 
================= 
grid :: <function> 
	Return the stellar mass-metallicity grid of fractional nucleosynethetic 
	yields for a given element and study to the user. 

References 
========== 
Karakas (2010), MNRAS, 403, 1413 
Cristallo (2011), ApJS, 197, 17 
"""

from __future__ import absolute_import
from .grid import yield_grid as grid
import sys 

__all__ = ["grid"]
__all__ = [str(i) for i in __all__] # appease python 2 strings 

del absolute_import 
if sys.version_info[0] < 3: 
	del i 
else:
	pass 
del sys 


