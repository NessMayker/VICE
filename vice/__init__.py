# This file, included with the VICE package, is protected under the terms of the 
# associated MIT License, and any use or redistribution of this file in original 
# or altered form is subject to the copyright terms therein. 

"""
VICE: Versatile Integrator for Chemical Evolution
=================================================
A software built for numerical integration of single-zone chemical evolution 
models. 

See LICENSE for copyright information and citation requirements. 

Documentation for this package is available in two forms: 
	1) In the docstrings of the objects and functions
	2) Under docs/ in VICE's git repository at: 
		<https://github.com/giganano/VICE/tree/master/docs>

We recommend first-time users visit the git repository to obtain copies of the 
User's Guide and Science Documentation. Under the docs/ directory, they will 
find a QuickStartTutorial.ipynb notebook intended for quick familiarization 
with the structure of VICE. 

It is also recommended that VICE users install the package dill, an extension 
of the python standard library package pickle. This enables VICE to encode 
functional attributes with its output, and can be achieved via pip. 

In all docstrings, examples of code are represented by three > signs:

	>>> a = 5
	>>> a += 10


Included Features
=================

The VICE Dataframe 
------------------
	A data storing object meant for case-insensitive lookup 

	Instances of this class included with VICE:
	-------------------------------------------
		atomic_number: 
			The number of protons in the nucles of each recognized element 
		yields.ccsne.settings: 
			User-specified yield settings from core collapse supernovae 
		yields.sneia.settings: 
			User-specified yield settings from type Ia supernovae 
		solar_z: 
			The solar abundance by mass of each element calibrated by Asplund 
			et al. (2009), ARA&A, 47, 481
		sources: 
			The dominant enrichment sources for each element 

Classes 
-------
	singlezone: 
		Run simulations of single-zone galactic chemical evolution models 
	output: 
		Handle the output of the integrator class 

Functions 
---------
	yields.agb.grid: 
		Returns a built-in mass-metallicity yield grid for a given element 
	yields.ccsne.fractional: 
		Returns an IMF-integrated yield for a given element from core-collapse 
		supernovae 
	yields.sneia.fractional: 
		Returns an IMF-integrated yield for a give nelement from type Ia 
		supernovae 
	history: 
		Read in simulation output containing the time-evolution of the ISM 
	mdf: 
		Read in the stellar metallicity distribution function from a simulation 
	mirror: 
		Given an output object, returns an integrator with the same properties 
		as that which produced the output 
	yields.sneia.single: 
		Returns the mass of a given element produced by a single instance of 
		a type Ia supernova on average 
	single_stellar_population: 
		Returns the mass of a given element produced over time by a single 
		episode of star formation

Command Line 
============
VICE also allows users to run simulations of simple evolutionary histories 
directly from the command line. Type "vice -h" in a terminal for instructions 
on how to do so. 
"""

from __future__ import absolute_import
import warnings 
import sys
import os

__author__ = "James W. Johnson <giganano9@gmail.com>"

try: 
	__VICE_SETUP__ 
except NameError:
	__VICE_SETUP__ = False

if __VICE_SETUP__: 
	from ._build_utils import *
	_LONG_DESCRIPTION_ = __doc__
else:
	try:
		from .version import version as __version__
		from .version import release as __release
	except: 
		message = "Error importing VICE. You should not try to run VICE from " 
		message += "within its source directory. Please exit the VICE source "
		message += "tree and relaunch your python interpreter from there." 
		raise ImportError(message)

	if not __release: 
		warnings.warn("Using un-released version of VICE", UserWarning)

	__all__ = ["__author__", "__version__", "yields"] 

	from .core import * 
	from ._build_utils import *
	from . import yields

	__all__.extend(core.__all__)
	__all__.extend(_build_utils.__all__)

	"""
	Remove locally imported variables that aren't needed, but leave the user 
	the ability to call vice.warnings to get only VICE to keep quiet and let 
	the rest of python raise warnings. 
	"""
	del core
	del _build_utils
	del version
	del __release
	del absolute_import
	del sys
	del os

