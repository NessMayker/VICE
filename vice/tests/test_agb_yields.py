
from __future__ import print_function 
from vice.yields.agb import grid 
from vice import _RECOGNIZED_ELEMENTS_ 
from vice import atomic_number 

_STUDY_ = ["cristallo11", "karakas10"] 

if __name__ == "__main__": 
	print("TESTING: vice.yields.agb.grid") 
	out = open("test_agb_yields.out", 'w') 
	for i in _STUDY_: 
		success = True 
		for j in _RECOGNIZED_ELEMENTS_: 
			if not (i == "karakas10" and atomic_number[j] > 28): 
				try: 
					foo = grid(j, study = i) 
				except: 
					success = False 
			else: 
				pass 
		if success: 
			message = "%s: Success" % (i) 
		else:
			message = "%s: Failed" % (i) 
		print(message) 
		out.write("%s\n" % (message)) 
	out.close() 





