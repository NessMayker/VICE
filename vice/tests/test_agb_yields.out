cristallo11: Success
karakas10: Success
